packages/mds-ingest-service/repository/index.ts

```diff
@@ -14,14 +14,226 @@
+  end: Timestamp
+}
+export interface GetVehicleEventsFilterParams {
+  vehicle_types?: VEHICLE_TYPE[]
+  propulsion_types?: PROPULSION_TYPE[]
+  provider_ids?: UUID[]
+  vehicle_states?: VEHICLE_STATE[]
+  time_range: TimeRange
+  grouping_type: GROUPING_TYPE
+  device_or_vehicle_id?: string // Match on device_id or vehicle_id
+  device_ids?: UUID[]
+  event_types?: VEHICLE_EVENT[]
+  geography_ids?: UUID[]
+}
 class IngestReadWriteRepository extends ReadWriteRepository {
   constructor() {
     super('ingest', { entities, migrations })
   }
+
+  public createEvents = async (events: EventDomainCreateModel[]) => {
+    const { connect } = this
+    try {
+      const connection = await connect('rw')
+      const { raw: entities }: InsertReturning<EventEntity> = await connection
+        .getRepository(EventEntity)
+        .createQueryBuilder()
+        .insert()
+        .values(events.map(EventDomainToEntityCreate.mapper()))
+        .returning('*')
+        .execute()
+      return entities.map(EventEntityToDomain.map)
+    } catch (error) {
+      throw RepositoryError(error)
+    }
+  }
+
+  public createTelemetries = async (events: TelemetryDomainCreateModel[]) => {
+    const { connect } = this
+    try {
+      const connection = await connect('rw')
+      const { raw: entities }: InsertReturning<TelemetryEntity> = await connection
+        .getRepository(TelemetryEntity)
+        .createQueryBuilder()
+        .insert()
+        .values(events.map(TelemetryDomainToEntityCreate.mapper()))
+        .returning('*')
+        .execute()
+      return entities.map(TelemetryEntityToDomain.map)
+    } catch (error) {
+      throw RepositoryError(error)
+    }
+  }
+
+  public createDevices = async (events: DeviceDomainCreateModel[]) => {
+    const { connect } = this
+    try {
+      const connection = await connect('rw')
+      const { raw: entities }: InsertReturning<DeviceEntity> = await connection
+        .getRepository(DeviceEntity)
+        .createQueryBuilder()
+        .insert()
+        .values(events.map(DeviceDomainToEntityCreate.mapper()))
+        .returning('*')
+        .execute()
+      return entities.map(DeviceEntityToDomain.map)
+    } catch (error) {
+      throw RepositoryError(error)
+    }
+  }
+
+  public getEvents = async (params: GetVehicleEventsFilterParams): Promise<EventDomainModel[]> => {
+    const { connect } = this
+    const {
+      time_range: { start, end },
+      // geography_ids,
+      grouping_type = 'latest_per_vehicle',
+      event_types,
+      vehicle_states,
+      vehicle_types,
+      device_or_vehicle_id,
+      device_ids,
+      propulsion_types,
+      provider_ids
+    } = params
+    try {
+      const connection = await connect('ro')
+
+      const query = connection
+        .getRepository(EventEntity)
+        .createQueryBuilder('events')
+        .innerJoin(qb => qb.from(DeviceEntity, 'd'), 'devices', 'devices.device_id = events.device_id')
+        .leftJoinAndMapOne(
```

Comment:
Are there indexes in place such that these joins will be efficient in the absence of any foreign keys?

Reply:
events table has a primary key of `"events_pkey" PRIMARY KEY, btree (device_id, "timestamp")`

Reply:
also, there's a new index on (trip_id, timestamp), for the "last by trip_id" join filter.